### PR TITLE
feat: reconfigure project

### PR DESCRIPTION
**Pull request by estrada9166**
- Closes  [UNIFY-1038](https://cypress-io.atlassian.net/browse/UNIFY-1038)

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts [hunk 1]
```diff
@@ -99,6 +99,13 @@ export const CurrentProject = objectType({
       },
     })
 
+    t.boolean('forceReconfigureProject', {
```
Comment (ZachJW34) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts:102]: Rather than surfacing an additional property, would it be better to modify `isCtConfigured` and `isE2eConfigured` to take into accound the `forceReconfigureProject` flag e.g.
 `isCtConfigured: { resolve: ctx.lifecycleManager.isTestingTypeConfigured('component') && !ctx.coreData.forceReconfigureProject }`.
 
 Also, do you know if the the project setup should be testingType specific? The user could have selected component testing, hit the reconfigureProject and then hit the back button in the launchpad project setup page and click on e2e testing which would bring them into e2e setup.

File: packages/data-context/src/actions/WizardActions.ts [hunk 2]
```diff
@@ -71,17 +71,19 @@ export class WizardActions {
 
     assert(currentTestingType)
     assert(chosenLanguage)
-    assert(!this.ctx.lifecycleManager.isTestingTypeConfigured(currentTestingType), `Cannot call this if the testing type (${currentTestingType}) has not been configured`)
+
     switch (currentTestingType) {
       case 'e2e': {
         this.ctx.coreData.scaffoldedFiles = await this.scaffoldE2E()
         this.ctx.lifecycleManager.refreshMetaState()
+        this.ctx.coreData.forceReconfigureProject = false
```
Comment (ZachJW34) [inline packages/data-context/src/actions/WizardActions.ts:79]: Minor: I'm seeing the `ctx.update` function used more and more. Is this something we should make standard and apply to all updates of coreData, including this update?

**Review by ZachJW34 — COMMENTED** [review]
Tested, works great just have a few questions

File: packages/data-context/src/actions/ProjectActions.ts [hunk 3]
```diff
@@ -467,8 +469,12 @@ export class ProjectActions {
     return { specs, specPattern, excludeSpecPattern, additionalIgnorePattern }
   }
 
-  async reconfigureProject () {
+  async reconfigureProject (forceReconfigureProject?: boolean) {
     // Initialize active project close first the current project
+    if (forceReconfigureProject) {
```
Comment (lmiller1990) [inline packages/data-context/src/actions/ProjectActions.ts:474]:
```ts
async reconfigureProject (forceReconfigureProject: boolean = false) {
 this.ctx.update(coreData => {
    coreData.forceReconfigureProject = forceReconfigureProject
  })
}
```

Comment (lmiller1990) [reply packages/data-context/src/actions/ProjectActions.ts:474]: Up to you

File: packages/data-context/src/actions/WizardActions.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/WizardActions.ts:79]: Yes we should! It took me a while to grok how this all is supposed to work, I started documenting it here. https://github.com/cypress-io/cypress/tree/10.0-release/packages/data-context#actions Please help if you see or know some better guidelines. Basically if everything can mutate via a single place, like `update`, we will theoretically be able to "play back" a session - good for debugging.

There are still some unknowns, but I'd like to get the workflow more formal.

Basically:

- coreData: the data
- sources: read/derive coreData, readonly
- actions: mutate data

Kind of like in Redux...

- state: immutable
- reducer: update state
- selector: derive state

File: packages/data-context/src/actions/WizardActions.ts [hunk 4]
```diff
@@ -71,17 +71,19 @@ export class WizardActions {
...
 
         return chosenLanguage
       }
       case 'component': {
         this.ctx.coreData.scaffoldedFiles = await this.scaffoldComponent()
         this.ctx.lifecycleManager.refreshMetaState()
+        this.ctx.coreData.forceReconfigureProject = false
```
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:86]:
```ts
this.ctx.update(coreData => {
  coreData.forceReconfigureProject = false
})
```

**Review by lmiller1990 — COMMENTED** [review]
Not sure about the `forceReconfigProject` flag, but maybe this is the cleanest way to do this.

Also, I think we should start making an effort to do `this.ctx.update` everywhere. My understanding is we should always update via `this.ctx.update` (just understood recently): https://github.com/cypress-io/cypress/tree/10.0-release/packages/data-context#actions
